modules/xmpp/JingleSessionPC.js

```diff
@@ -969,155 +1160,166 @@ export default class JingleSessionPC extends JingleSession {
-        //  core flows like this.
         return new Promise((resolve, reject) => {
-            const remoteUfrag = JingleSessionPC.getUfrag(remoteDescription.sdp);
+            if (this.peerconnection.signalingState === 'closed') {
+                reject('Attempted to renegotiate in state closed');
 
-            if (remoteUfrag !== this.remoteUfrag) {
-                this.remoteUfrag = remoteUfrag;
-                this.room.eventEmitter.emit(
-                        XMPPEvents.REMOTE_UFRAG_CHANGED, remoteUfrag);
+                return;
+            }
+            if (this.isInitiator) {
+                this._initiatorRenegotiate(remoteDescription, resolve, reject);
+            } else {
+                this._responderRenegotiate(remoteDescription, resolve, reject);
             }
+        });
+    }
+
+    /**
+     * Renegotiate cycle implementation for the responder case.
+     * @param {object} remoteDescription the SDP object as defined by the WebRTC
+     * which will be used as remote description in the cycle.
+     * @param {function} resolve the success callback
+     * @param {function} reject the failure callback
+     * @private
+     */
+    _responderRenegotiate(remoteDescription, resolve, reject) {
 
-            logger.debug('Renegotiate: setting remote description');
+        // FIXME use WebRTC promise API to simplify things
+        logger.debug('Renegotiate: setting remote description');
+        this.peerconnection.setRemoteDescription(
+            remoteDescription,
+            () => {
+                logger.debug('Renegotiate: creating answer');
+                this.peerconnection.createAnswer(
+                    answer => {
+                        logger.debug('Renegotiate: setting local description');
+                        this.peerconnection.setLocalDescription(
+                            answer,
+                            () => {
+                                this._dequeIceCandidates();
+                                resolve();
+                            },
+                            error => {
+                                reject(
+                                    `setLocalDescription failed: ${error}`);
+                            }
+                        );
+                    },
+                    error => reject(`createAnswer failed: ${error}`),
+                    this.mediaConstraints
+                );
+            },
+            error => reject(`setRemoteDescription failed: ${error}`)
+        );
+    }
+
+    /**
+     * Renegotiate cycle implementation for the initiator's case.
+     * @param {object} remoteDescription the SDP object as defined by the WebRTC
+     * which will be used as remote description in the cycle.
+     * @param {function} resolve the success callback
+     * @param {function} reject the failure callback
+     * @private
+     * @private
+     */
+    _initiatorRenegotiate(remoteDescription, resolve, reject) {
```

Comment:
I've tried, but it's easier said than done :P Probably promises could help. I have left a FIXME note